### PR TITLE
Implement Missing Property of Type `this`

### LINKED ISSUE #14104: Implement missing members incorrectly implements type of 'this'
```ts
abstract class C {
    abstract myMethod(): this;
}

class D/**/ extends C {

}
```

Try to apply a codefix to implement missing members of `C`.

**Expected**: a method is generated with the following text:

```ts
myMethod(): this {
    // ...
}
```

**Actual**

```ts
myMethod(): C {
    // ...
}
```

### PR DESCRIPTION
**Pull request by aozgaa**
Fixes #14104

### CODE REVIEW

File: src/services/codefixes/helpers.ts [hunk 1]
```diff
@@ -34,12 +32,22 @@ namespace ts.codefix {
...
             case SyntaxKind.PropertySignature:
             case SyntaxKind.PropertyDeclaration:
-                const typeString = checker.typeToString(type, enclosingDeclaration, TypeFormatFlags.None);
+                let typeString: string | undefined = undefined;
+                const typeAtOldDeclaration = checker.getTypeAtLocation(declaration);
+                if ((typeAtOldDeclaration as TypeParameter).isThisType) {
+                    typeString = "this";
+                }
+                else {
+                    typeString = checker.typeToString(typeAtNewDeclaration, enclosingDeclaration, TypeFormatFlags.None);
```
Comment (DanielRosenwasser) [inline src/services/codefixes/helpers.ts]: You could also consider another `TypeFormatFlags` member named `PreserveThisTypes`

**Comment by DanielRosenwasser** [comment]
Is this a work in progress? I don't see a test for the original case that I posted.

**Comment by aozgaa** [comment]
Can you take a look, @mhegazy, @DanielRosenwasser ?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -21106,7 +21106,15 @@ namespace ts {
             }
 
             if (isPartOfTypeNode(node)) {
-                return getTypeFromTypeNode(<TypeNode>node);
+                let typeFromTypeNode = getTypeFromTypeNode(<TypeNode>node);
+
+                if (typeFromTypeNode && isExpressionWithTypeArgumentsInClassImplementsClause(node)) {
```
Comment (aozgaa) [inline src/compiler/checker.ts]: Is this the right thing to check for to find out if we are in an `implements` clause?

Comment (aozgaa) [inline src/compiler/checker.ts]: Is this the right place to do the this instantiation or should I moce this logic into `getTypeFromTypeNode`?

I created a circularity (ie: stack overflow) when I moved it even further into the checker (namely by doing the instantiation at `getTypeFromTypeReference`).

Comment (mhegazy) [inline src/compiler/checker.ts]: you can use `isHeritageClauseElementIdentifier`

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -21106,7 +21106,15 @@ namespace ts {
             }
 
             if (isPartOfTypeNode(node)) {
-                return getTypeFromTypeNode(<TypeNode>node);
+                let typeFromTypeNode = getTypeFromTypeNode(<TypeNode>node);
+
+                if (typeFromTypeNode && isExpressionWithTypeArgumentsInClassImplementsClause(node)) {
+                    const containingClass = getContainingClass(node);
+                    const classType = getTypeOfNode(containingClass) as InterfaceType;
+                    typeFromTypeNode = getTypeWithThisArgument(typeFromTypeNode, classType.thisType);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: consider refactoring to have one check higher up, and then handle extends/implements accordinglly

**Comment by mhegazy** [comment]
```ts
            if (isHeritageClauseElementIdentifier(node)) {
                const expressionWithTypearguments = <ExpressionWithTypeArguments>getAncestor(node, SyntaxKind.ExpressionWithTypeArguments);
                const classDeclaration = tryGetClassExtendingExpressionWithTypeArguments(expressionWithTypearguments);
                if (classDeclaration) {
                    // A SyntaxKind.ExpressionWithTypeArguments is considered a type node, except when it occurs in the
                    // extends clause of a class. We handle that case here.
                    return getBaseTypes(<InterfaceType>getDeclaredTypeOfSymbol(getSymbolOfNode(classDeclaration)))[0];
                }
                else {
                    return getTypeFromTypeNode(<TypeNode>node);
                }
            }
```

**Comment by aozgaa** [comment]
It doesn't look like we can consolidates the checks as we discussed offline because the call to `isPartOfExpression` (above the original `isExpressionWithTypeArgumentsInClassExtendsClause` call) captures the case where the node is the LHS expression of of an `ExpressionWithTypeArguments`.*
Is this the desired behavior?

If so, we can't do the `isHeritageClauseElementIdentifier` check too early.

\* One test case that failed is when we have `class C extends null {...}` because with the change the type we get for `null` is undefined whereas before it was the null type.

In general, if we have `class D extends A.B.C<T> {...}`, if the node we are given is 
1) `A`,
2) `B`,
3) `C`,
4)  `A.B`,
5)  `A.B.C`,
6) `A.B.C<T>`,

in which cases do we want to instantiate the `this` type? My intuition is that we should only do it in case (6) and possibly also (3) and/or (5)

**Pull request merged**